Assets/Labrys/FeatureEditor/FeatureAsset.cs

```diff
@@ -0,0 +1,372 @@
+﻿using Labrys.Generation;
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Labrys.FeatureEditor
+{
+	/// <summary>
+	/// A file-backed representation of a Feature for use with the Feature Editor
+	/// </summary>
+	[CreateAssetMenu(menuName = "Labrys/Feature", fileName = "New Feature")]
+	public class FeatureAsset : ScriptableObject, ISerializationCallbackReceiver
+	{
+		[Serializable]
+		public class Section
+		{
+			public static bool IsValidPosition(Vector2Int gridPosition)
+			{
+				return gridPosition.x % GRID_DENSITY == 0 && gridPosition.y % GRID_DENSITY == 0;
+			}
+
+			public string Variant { get; set; }
+		}
+
+		[Serializable]
+		public class Link
+		{
+			public static void GetMinMaxSubjectTileCount(Vector2Int gridPos, out int min, out int max)
+			{
+				if (gridPos.x % 2 != 0)
+				{
+					if (gridPos.y % 2 != 0)
+					{
+						min = max = 4;
+					}
+					else
+					{
+						min = 1;
+						max = 2;
+					}
+				}
+				else if (gridPos.y % 2 != 0)
+				{
+					min = 1;
+					max = 2;
+				}
+				else
+				{
+					min = max = -1;
+				}
+			}
+
+			public static IEnumerable<Vector2Int> GetSubjectTileGridPositions(Vector2Int gridPos)
+			{
+				HashSet<Vector2Int> uniquePositions = new HashSet<Vector2Int>();
+
+				int lowerX = Mathf.FloorToInt(gridPos.x / GRID_DENSITY) * (int)GRID_DENSITY;
+				int upperX = Mathf.FloorToInt((gridPos.x / GRID_DENSITY) + 0.5f) * (int)GRID_DENSITY;
+				int lowerY = Mathf.FloorToInt(gridPos.y / GRID_DENSITY) * (int)GRID_DENSITY;
+				int upperY = Mathf.FloorToInt((gridPos.y / GRID_DENSITY) + 0.5f) * (int)GRID_DENSITY;
+
+				uniquePositions.Add(new Vector2Int(lowerX, lowerY));
+				uniquePositions.Add(new Vector2Int(lowerX, upperY));
+				uniquePositions.Add(new Vector2Int(upperX, lowerY));
+				uniquePositions.Add(new Vector2Int(upperX, upperY));
+
+				Vector2Int[] finalPositions = new Vector2Int[uniquePositions.Count];
+				uniquePositions.CopyTo(finalPositions);
+				return finalPositions;
+			}
+
+			public bool Open { get; set; } 
+			public bool External { get; set; }
+		}
+
+		public const float GRID_DENSITY = 2f;
+
+		private static Vector2Int[] dirVectors = new Vector2Int[] {
+				Vector2Int.right,
+				Vector2Int.up + Vector2Int.right,
+				Vector2Int.up,
+				Vector2Int.up + Vector2Int.left,
+				Vector2Int.left,
+				Vector2Int.down + Vector2Int.left,
+				Vector2Int.down,
+				Vector2Int.down + Vector2Int.right
+		};
+
+		private static Connection[] dirConnections = new Connection[] {
+				Connection.East,
+				Connection.Northeast,
+				Connection.North,
+				Connection.Northwest,
+				Connection.West,
+				Connection.Southwest,
+				Connection.South,
+				Connection.Southeast
+		};
+
+#if UNITY_EDITOR
```

Comment:
Why this macro?

Reply:
To prevent the creation of assets in production.  This method should only be used in the editor, and not in a standalone executable.

Attempts to use this method in standalone code will result in a compiler error when attempting to build a standalone.

Reply:
Do we envision this being a standalone? This is designed to be a project specifically for Unity